### PR TITLE
disttask: add operator abstraction

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #41495 

Problem Summary:

#### Operator
Operator can run business logic with multiple workers.
Note that we use workerPool to manage all workers.
Workers will handle task when there is a task send to it.
#### Data
Abstract DataSource and DataSink to handle the dataflow.
#### Pipeline
Pipeline is a chain of Operators.

The workflow is:
1. add task to the first operator, then this operator start to handle tasks.
2. When the task is handled, the operator send task to the next operator.

Moreover, we can read dataSource to add task. It will be implemented soon.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: disttask/framework/operator/asyncPipeline.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
+// Copyright 2023 PingCAP, Inc.
```
Comment (hawkingrei) [inline disttask/framework/operator/asyncPipeline.go:1]: ```async_pipeline.go```?

Comment (ywqzzy) [reply disttask/framework/operator/asyncPipeline.go:1]: ok, needs more refinement.

File: disttask/framework/operator/async_test.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	impl1 := NewAsyncOperatorImpl[asyncChunk]("impl1", newExampleAsyncOperatorImpl, poolutil.DDL, 10)
+	impl2 := NewAsyncOperatorImpl[asyncChunk]("impl2", newExampleAsyncOperatorImpl, poolutil.DDL, 10)
+	source0 := impl0.getSource()
+	source1 := impl1.getSource()
+	source2 := impl2.getSource()
+	sink := &simpleAsyncDataSink{0, 0, sync.Mutex{}}
+
+	impl0.setSink(source1.(DataSink))
+	impl1.setSink(source2.(DataSink))
+	impl2.setSink(sink)
```
Comment (D3Hunter) [inline disttask/framework/operator/async_test.go:36]: if we're sure pipeline is a chain, better provide a method in pipeline to connect them in an easy way

File: disttask/framework/operator/data.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+// limitations under the License.
+
+package operator
+
+// DataSource defines interface for source of all operators.
+// Operator read data from DataSource, then process the read data.
+type DataSource interface {
+	HasNext() bool
+	Read() (any, error)
+	Display() string
```
[the remark this answers is not in the document]
Comment (ywqzzy) [reply disttask/framework/operator/data.go:22]: > what's the purpose of this?

For better tracking of the operator chain.

File: disttask/framework/operator/data.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operator
+
+// DataSource defines interface for source of all operators.
+// Operator read data from DataSource, then process the read data.
+type DataSource interface {
+	HasNext() bool
```
[the remark this answers is not in the document]
Comment (ywqzzy) [reply disttask/framework/operator/data.go:20]: > we might need a Start method to do some prepare work before uses this source. `HasNext` can be merged with `Read` to be `Next`, we can return EOF when no data.

Good point. I will add it.

File: disttask/framework/operator/async_test.go (hunk 2 above)
Comment (ywqzzy) [reply disttask/framework/operator/async_test.go:36]: > if we're sure pipeline is a chain, better provide a method in pipeline to connect them in an easy way

Yep, I will try.

File: disttask/framework/operator/async_pipeline.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operator
+
+// AsyncPipeline wraps a list of AsyncOperatorImpls.
+// The dataflow is from the first operator to the last operator.
+//
+//	Eg: op1.AddTask ---> op1.HandleTask  ---> op2.AddTask ---> op2.HandleTask
```
[the remark this answers is not in the document]
Comment (ywqzzy) [reply disttask/framework/operator/async_pipeline.go:20]: > why have `task` in operator?
Every operator using a workerpool to start workers.
We need task to trigger workers to run tasks.

File: disttask/framework/operator/async.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+import (
+	"github.com/pingcap/tidb/resourcemanager/pool/workerpool"
+	poolutil "github.com/pingcap/tidb/resourcemanager/util"
+)
+
+// AsyncOperator provide basic getter setter for each operator.
+type AsyncOperator[T any] struct {
+	source DataSource
+	sink   DataSink
+	pool   *workerpool.WorkerPool[T] // workers running on pool
```
[the remark this answers is not in the document]
Comment (ywqzzy) [reply disttask/framework/operator/async.go:26]: > this `pool` shouldn't be part of a operator, the pipeline that execute those operators should use it.

Ok, pipeline hold the pools will be cleaner.

File: disttask/framework/operator/async_test.go (hunk 2 above)
Comment (ywqzzy) [reply disttask/framework/operator/async_test.go]: > if we're sure pipeline is a chain, better provide a method in pipeline to connect them in an easy way

I will do it when plan is supported.

File: disttask/framework/operator/async.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ywqzzy) [reply disttask/framework/operator/async.go]: > > this `pool` shouldn't be part of a operator, the pipeline that execute those operators should use it.
> 
> Ok, pipeline hold the pools will be cleaner.

I will do it when we support better way to build pipeline with plan.

File: disttask/framework/operator/data.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operator
+
+// DataSource defines interface for source of all operators.
+// Operator read data from DataSource, then process the read data.
+type DataSource[T any] interface {
+	Start() error
+	Next() (T, error)
```
Comment (tangenta) [inline disttask/framework/operator/data.go:21]: How about renaming it to `Read()`? It should be complementary to `DataSink.Write()`.

Comment (ywqzzy) [reply disttask/framework/operator/data.go:21]: > How about renaming it to `Read()`? It should be complementary to `DataSink.Write()`.

Ok.

**Pull request closed without merging**